File: sentinelai-frontend/src/types/ioc.ts
export interface IocEntry {
  id: string;
  ioc_type: string;
  ioc_value: string;
  normalized_value: string;
  confidence: number;
  source_event: string | null;
  source_log: string | null;
  source_ip: string | null;
  first_seen: string;
  last_seen: string;
  occurrences: number;
  severity: string;
  status: string;
  tags: Record<string, unknown> | null;
  metadata: Record<string, unknown> | null;
  source_ids: string[] | null;
  context: string | null;
  kill_chain_phase: string | null;
  created_at: string;
  updated_at: string;
}

export interface IocStats {
  total: number;
  by_type: Record<string, number>;
  by_severity: Record<string, number>;
  by_status: Record<string, number>;
  top_source_ips: { source_ip: string; count: number }[];
  latest_iocs: { id: string; ioc_type: string; ioc_value: string; severity: string; last_seen: string }[];
  unique_domains: number;
  unique_ips: number;
  unique_hashes: number;
}

export interface IocExtractResult {
  extracted: number;
  new: number;
  updated: number;
  iocs: IocEntry[];
}

export const IOC_TYPE_MAP: Record<string, { label: string; color: string }> = {
  ipv4: { label: 'IPv4', color: '#3b82f6' },
  ipv6: { label: 'IPv6', color: '#6366f1' },
  domain: { label: 'Domain', color: '#10b981' },
  url: { label: 'URL', color: '#f59e0b' },
  hostname: { label: 'Hostname', color: '#14b8a6' },
  email: { label: 'Email', color: '#8b5cf6' },
  username: { label: 'Username', color: '#ec4899' },
  md5: { label: 'MD5', color: '#ef4444' },
  sha1: { label: 'SHA1', color: '#f97316' },
  sha256: { label: 'SHA256', color: '#dc2626' },
  registry_key: { label: 'Registry Key', color: '#06b6d4' },
  windows_sid: { label: 'Windows SID', color: '#84cc16' },
  process_name: { label: 'Process', color: '#22c55e' },
  executable_path: { label: 'Executable Path', color: '#eab308' },
  command_line: { label: 'Command Line', color: '#a855f7' },
  cve: { label: 'CVE', color: '#b91c1c' },
  mitre_technique: { label: 'MITRE Technique', color: '#7c3aed' },
  port: { label: 'Port', color: '#0ea5e9' },
  protocol: { label: 'Protocol', color: '#2dd4bf' },
};

export const IOC_SEVERITY_COLORS: Record<string, string> = {
  critical: 'text-red-400 bg-red-400/10 border-red-400/30',
  high: 'text-orange-400 bg-orange-400/10 border-orange-400/30',
  medium: 'text-yellow-400 bg-yellow-400/10 border-yellow-400/30',
  low: 'text-green-400 bg-green-400/10 border-green-400/30',
};

File: sentinelai-backend/app/schemas/ioc.py
from datetime import datetime
from typing import Any

from pydantic import Field

from app.schemas.base import BaseSchema


class IocEntryResponse(BaseSchema):
    id: str
    ioc_type: str
    ioc_value: str
    normalized_value: str
    confidence: float
    source_event: str | None = None
    source_log: str | None = None
    source_ip: str | None = None
    first_seen: datetime
    last_seen: datetime
    occurrences: int
    severity: str
    status: str
    tags: dict[str, Any] | None = None
    metadata: dict[str, Any] | None = Field(None, validation_alias="extra_data")
    source_ids: list[str] | None = None
    context: str | None = None
    kill_chain_phase: str | None = None
    created_at: datetime
    updated_at: datetime


class IocStatsResponse(BaseSchema):
    total: int
    by_type: dict[str, int]
    by_severity: dict[str, int]
    by_status: dict[str, int]
    top_source_ips: list[dict[str, Any]]
    latest_iocs: list[dict[str, Any]]
    unique_domains: int
    unique_ips: int
    unique_hashes: int


class IocExtractResponse(BaseSchema):
    extracted: int
    new: int
    updated: int
    iocs: list[IocEntryResponse]


class IocSearchParams(BaseSchema):
    q: str | None = None
    ioc_type: str | None = None
    severity: str | None = None
    status: str | None = None
    source_ip: str | None = None
    source_log: str | None = None
    date_from: datetime | None = None
    date_to: datetime | None = None
    page: int = 1
    page_size: int = 20
    sort_by: str = "last_seen"
    sort_order: str = "desc"
